### PR TITLE
COZMO-7520 Add Pet detection support

### PR DESCRIPTION
**Pull request by MarkWesleyAtAnki**
Added pets module, similar to faces, along with annotation support.
Cozmo can now detect cat and dog faces, and these can be used mostly the same as normal faces (except cat and dog faces don't provide 3D pose information, the faces aren't recognized between sightings (e.g. face leaving screen and coming back appears as a new pet) and therefore cannot be stored or named either).

### CODE REVIEW

File: src/cozmo/pets.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+The :class:`cozmo.world.World` object keeps track of pets the robot currently
+knows about, along with those that are currently visible to the camera.
+
+Each pet is assigned a :class:`Pet` object, which generates a number of
+observable events whenever the pet is observed, etc.
+
+If a pet goes off-screen, it will be assigned a new object_id (and
+therefore Pet object) when it returns. This is because the system can
+only tell if something appears to be a cat or a dog; it cannot recognise
```
Comment (msintov) [inline src/cozmo/pets.py]: "recognise" should be "recognize"

File: src/cozmo/pets.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+a specific pet, or e.g. tell the difference between two dogs.
```
Comment (msintov) [inline src/cozmo/pets.py]: I would change "or e.g. tell" to "or, for instance, tell".

File: src/cozmo/pets.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+Note that these pet-specific events are also passed up to the
+:class:`cozmo.world.World` object, so events for all known pets can be
```
Comment (msintov) [inline src/cozmo/pets.py]: I think "all known pets" should be "all pets" or "all recognized pets" perhaps?

File: src/cozmo/pets.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    '''
+    pet = 'The Pet instance that was observed'
+    updated = 'A set of field names that have changed'
+    image_box = 'A comzo.util.ImageBox defining where the pet is within Cozmo\'s camera view'
+
+
+class EvtPetAppeared(event.Event):
+    '''Triggered whenever an object is first visually identified by a robot.
+
+    This differs from Evt{etObserved in that it's only triggered when
```
Comment (msintov) [inline src/cozmo/pets.py]: "Evt{etObserved" should be "EvtPetObserved"

File: src/cozmo/pets.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    elif clad_pet_type == _clad_to_game_anki.Vision.PetType.Cat:
+        return PET_TYPE_CAT
+    elif clad_pet_type == _clad_to_game_anki.Vision.PetType.Dog:
+        return PET_TYPE_DOG
+    else:
+        raise ValueError("Unexpected pet type %s" % clad_pet_type)
+
+
+class Pet(event.Dispatcher):
+    '''A single pet (face) that Cozmo has detected.
```
Comment (msintov) [inline src/cozmo/pets.py]: Maybe remove "(face)"? Most or all other references to pet just say "pet", not "pet's face" or similar.

File: src/cozmo/pets.py (hunk 4 above)
Comment (gwatts) [inline src/cozmo/pets.py]: typo

File: src/cozmo/pets.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    See EvtPetAppeared if you only want to know when a pet first
+    becomes visible.
+    '''
+    pet = 'The Pet instance that was observed'
+    updated = 'A set of field names that have changed'
+    image_box = 'A comzo.util.ImageBox defining where the pet is within Cozmo\'s camera view'
+
+
+class EvtPetAppeared(event.Event):
+    '''Triggered whenever an object is first visually identified by a robot.
```
Comment (gwatts) [inline src/cozmo/pets.py]: s/object/pet/

File: src/cozmo/pets.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+#: assuming that Cozmo can no longer see a pet.
+PET_VISIBILITY_TIMEOUT = objects.OBJECT_VISIBILITY_TIMEOUT
+
+# Pet types that Cozmo can distinguish
+#: Pet Type reported by Cozmo when unsure of type of pet
+PET_TYPE_UNKOWN = "Unknown"
+#: Pet Type reported by Cozmo when he thinks it's a cat
+PET_TYPE_CAT = "Cat"
+#: Pet Type reported by Cozmo when he thinks it's a dog
+PET_TYPE_DOG = "Dog"
```
Comment (gwatts) [inline src/cozmo/pets.py]: nit pick, but can we make the constant values lowercase so there's one less thing to think about? e.g. `PET_TYPE_DOG = "dog"`

File: src/cozmo/pets.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+            return math.inf
+        return time.time() - self.last_observed_time
+
+    @property
+    def is_visible(self):
+        '''bool: True if the pet has been observed recently.
+
+        "recently" is defined as :const:`PET_VISIBILITY_TIMEOUT` seconds.
+        '''
+        return self._is_visible
```
Comment (gwatts) [inline src/cozmo/pets.py]: not liking that we have a bunch of essentially duplicate code now :-(

File: src/cozmo/world.py [hunk 9]
```diff
@@ -291,6 +325,14 @@ def _recv_msg_robot_erased_enrolled_face(self, evt, *, msg):
         if face:
             face.dispatch_event(evt)
 
+    def _recv_msg_robot_observed_pet(self, evt, *, msg):
+        pet = self._pets.get(msg.petID)
+        if not pet:
+            pet = self._allocate_pet_from_msg(msg)
+            print("alloc pet %s" % pet)
```
Comment (gwatts) [inline src/cozmo/world.py]: stray debug print

**Comment by MarkWesleyAtAnki** [comment]
@gwatts We could move some of this logic down to objects.ObservableObject and have faces.Face and pets.Pet extend that (in the same way that objects.LightCube and objects.Charger do), although e.g. Pet doesn't have a pose and . Or, probably better, add a new objects.ObservableThing (ObservableBase? Observable?) and move just the common elements down to that

**Pull request merged**